### PR TITLE
now compatible for both Cloudstack and Cosmic

### PR DESCRIPTION
**Pull request by ddegoede**
- Added Cosmic check for choosing whether or not to detach a disk before updating a disk.
- Fixed all tests.

For Cosmic the gateway parameter for networks is not required anymore, if the cidr contains a host address instead of the network address. But with the gateway parameter it works in both Cloudstack and Cosmic.

The following resources will not work on Cloudstack:

- Egress_Firewall: API response not compatible
- Firewall: API response not compatible
- Private_Gateway: API is not compatible
- Static_Route: API is not compatible

### CODE REVIEW

File: cosmic/resource_cosmic_loadbalancer_rule_test.go [hunk 1]
```diff
@@ -416,8 +419,9 @@ resource "cosmic_loadbalancer_rule" "foo" {
 	CLOUDSTACK_VPC_OFFERING,
 	CLOUDSTACK_ZONE,
 	CLOUDSTACK_VPC_NETWORK_CIDR,
+	CLOUDSTACK_VPC_NETWORK_GATEWAY,
 	CLOUDSTACK_VPC_NETWORK_OFFERING,
 	CLOUDSTACK_SERVICE_OFFERING_1,
 	CLOUDSTACK_TEMPLATE,
 	CLOUDSTACK_SERVICE_OFFERING_1,
-	CLOUDSTACK_TEMPLATE)
+	CLOUDSTACK_TEMPLATE)
```
Comment (svanharmelen) [inline cosmic/resource_cosmic_loadbalancer_rule_test.go]: Can you add a newline char here?

File: cosmic/resource_cosmic_egress_firewall_test.go [hunk 2]
```diff
@@ -73,24 +73,24 @@ func TestAccCosmicEgressFirewall_update(t *testing.T) {
...
 						CLOUDSTACK_NETWORK_1_IPADDRESS1+"/32"),
 					resource.TestCheckResourceAttr(
-						"cosmic_egress_firewall.foo", "rule.739924765.protocol", "tcp"),
+						"cosmic_egress_firewall.foo", "rule.1790767400.protocol", "tcp"),
 					resource.TestCheckResourceAttr(
-						"cosmic_egress_firewall.foo", "rule.739924765.ports.1889509032", "80"),
+						"cosmic_egress_firewall.foo", "rule.1790767400.ports.1889509032", "80"),
 				),
 			},
 		},
```
Comment (svanharmelen) [inline cosmic/resource_cosmic_egress_firewall_test.go]: Can you undo the changes in this file? As you mentioned yourself, these are probably related to the input so for now I would like to leave them unchanged.

File: cosmic/resource_cosmic_network_test.go [hunk 3]
```diff
@@ -275,4 +277,5 @@ resource "cosmic_network" "foo" {
 	CLOUDSTACK_VPC_OFFERING,
 	CLOUDSTACK_ZONE,
 	CLOUDSTACK_VPC_NETWORK_CIDR,
-	CLOUDSTACK_VPC_NETWORK_OFFERING)
+	CLOUDSTACK_VPC_NETWORK_GATEWAY,
+	CLOUDSTACK_VPC_NETWORK_OFFERING)
```
Comment (svanharmelen) [inline cosmic/resource_cosmic_network_test.go]: Again, please add a newline...

File: cosmic/resource_cosmic_disk.go [hunk 4]
```diff
@@ -370,3 +372,13 @@ func retryableAttachVolumeFunc(
...
+
+func isCosmic(cs *cosmic.CosmicClient) (bool) {
+	l := cs.Configuration.NewListCapabilitiesParams()
+	c, err  := cs.Configuration.ListCapabilities(l)
+	if err != nil {
+		fmt.Errorf("Unable to retrieve capabilities: %s", err)
+		return false
+	}
+	return c.Capabilities.Cosmic
+}
```
Comment (svanharmelen) [inline cosmic/resource_cosmic_disk.go]: Please move this func to the `resources.go` file.

File: cosmic/resource_cosmic_egress_firewall_test.go (hunk 2 above)
Comment (ddegoede) [inline cosmic/resource_cosmic_egress_firewall_test.go]: I left this in as these changes are required to test against Cosmic.

**Comment by svanharmelen** [comment]
Thanks @ddegoede!
